### PR TITLE
Add interactive tests for emacs mode

### PR DESCRIPTION
**Pull request by krader1961**
This is based on PR#788 by @siteshwar but with extensive changes to make
the tests more robust. In particular this structure of the unit test makes
it easier to add a new test. It also introduces a couple of helper
functions to make reading the log files easier.

### CODE REVIEW

File: src/cmd/ksh93/tests/emacs.exp [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+expect "\r\nbuh bye\r\n" {
+    puts "echo 2 works"
+}
+expect_prompt
+
+# ==========
+# Send ctrl-p to retrieve previous command
+log_test_entry
+send [ctrl P]
+send [ctrl L]
```
Comment (siteshwar) [inline src/cmd/ksh93/tests/emacs.exp]: Why are you sending `ctrl-L` here ?

Comment (siteshwar) [reply src/cmd/ksh93/tests/emacs.exp]: We can probably wrap this control sequence in a procedure called `redraw_prompt` for better readability.

**Comment by siteshwar** [comment]
There was a timing issue in my original test, how was it fixed ?

File: src/cmd/ksh93/tests/emacs.exp (hunk 1 above)
Comment (krader1961) [reply src/cmd/ksh93/tests/emacs.exp]: Okay. Will do.

**Comment by krader1961** [comment]
> There was a timing issue in my original test, how was it fixed ?

I can't answer that specific question because of the significant restructuring did. But I can answer the more general question. Part of the challenge to writing these types of tests is ensuring the tests are in sync with the buffered output. This is why many of the tests check for `: some text` or `\r\nsome text\r\n`. The first pattern is used after sending [ctrl-L] to redraw the command line and helps ensure we are matching the command line by anchoring it to the last two characters of the prompt (the `: `). The second pattern is bracketed by `\r\n` to ensure it matches the output of executing the `echo` or whatever command was run rather than the same text on the command line. Keep in mind that such matches implicitly consume any prior characters in the buffer.

**Pull request closed without merging**